src/plonk/permutation/verifier.rs

```diff
@@ -76,87 +104,145 @@ impl<C: CurveAffine> Evaluated<C> {
         vk: &'a plonk::VerifyingKey<C>,
         p: &'a Argument,
         advice_evals: &'a [C::Scalar],
-        fixed_evals: &[C::Scalar],
+        fixed_evals: &'a [C::Scalar],
         instance_evals: &'a [C::Scalar],
         l_0: C::Scalar,
+        l_last: C::Scalar,
+        l_blind: C::Scalar,
         beta: ChallengeBeta<C>,
         gamma: ChallengeGamma<C>,
         x: ChallengeX<C>,
     ) -> impl Iterator<Item = C::Scalar> + 'a {
+        let chunk_len = vk.cs.degree() - 2;
         iter::empty()
-            // l_0(X) * (1 - z(X)) = 0
-            .chain(Some(
-                l_0 * &(C::Scalar::one() - &self.permutation_product_eval),
-            ))
-            // z(X) \prod (p(X) + \beta s_i(X) + \gamma)
-            // - z(omega^{-1} X) \prod (p(X) + \delta^i \beta X + \gamma)
-            .chain(Some({
-                let mut left = self.permutation_product_eval;
-                for (eval, permutation_eval) in p
-                    .columns
+            // Enforce only for the first set.
```

Comment:
Is it possible to reduce the duplication with the prover? (Does not block.)

Reply:
It's too challenging for this PR, but at least theoretically possible for some of the expressions. We should explore this.